spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java

```diff
@@ -407,14 +422,35 @@ public Scan build() {
 
   private Scan buildBatchScan() {
     Schema expectedSchema = schemaWithMetadataColumns();
+    org.apache.iceberg.Scan scan =
+        buildIcebergBatchScan(false /* not include Column Stats */, expectedSchema);
+    if (pushedLimit > 0 && hasDeletes(scan)) {
+      LOG.info("Skipping limit pushdown: detected row level deletes");
+      pushedLimit = -1;
+    }
     return new SparkBatchQueryScan(
         spark,
         table,
-        buildIcebergBatchScan(false /* not include Column Stats */, expectedSchema),
+        scan,
         readConf,
         expectedSchema,
         filterExpressions,
-        metricsReporter::scanReport);
+        metricsReporter::scanReport,
+        pushedLimit);
+  }
+
+  private boolean hasDeletes(org.apache.iceberg.Scan scan) {
+    try (CloseableIterable<FileScanTask> fileScanTasks = scan.planFiles()) {
```

Comment:
We are now potentially scanning the file scan tasks several times before beginning a query, 
1. Push aggregates
2. Has Deletes
3. Actually creating tasks

I'm wondering if we should be caching this

Reply:
Probably save this for another PR

Reply:
I will have a follow-up PR for this.